### PR TITLE
[megatron] fix: patch support newer mcore version

### PR DESCRIPTION
**Pull request by HollowMan6**
### What does this PR do?

Patch support newer mcore version

Tested on https://github.com/NVIDIA/Megatron-LM/commit/bbbedbb9f53343762e4dc70abc771b813a83d817

### Checklist Before Starting

- [X] Search for similar PRs. Paste at least one query link here: ...
- [X] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [X] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [X] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [X] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [X] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [X] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [X] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/models/mcore/patch.py [hunk 1]
```diff
@@ -258,13 +258,14 @@ def patch_forward(
...
         # query: [96, 1, 16, 128], key:[96, 1, 16, 128], value:[96, 1, 16, 128]
-        query, key, value = self.get_query_key_value_tensors(
+        qkv = self.get_query_key_value_tensors(
             hidden_states,
             key_value_states,
             position_ids,
             packed_seq_params,
             inference_context=inference_context,
         )
+        query, key, value = qkv[:3]
```
Comment (vermouth1992) [inline verl/models/mcore/patch.py:268]: What's the meaning of this change?

Comment (HollowMan6) [reply verl/models/mcore/patch.py:268]: In https://github.com/NVIDIA/Megatron-LM/commit/382eeea3951a6a333e023689e4b9e2f568f23e41#diff-d7e453d550e7f78221ea3445018264015c1e83c0db892a41253bde71d069a202L247, the return value for `get_query_key_value_tensors` has changed from `query, key, value` to `query, key, value, q_compressed, kv_compressed`. `kv_compressed` is not used in the forward pass so it just gets ignored, here I write in this way to take the first 3 returned values for backward compatibility.

I just found that `q_compressed` is actually used for DSA like DeepSeek V3.2, so I made and pushed a modification to fix it.

Comment (vermouth1992) [reply verl/models/mcore/patch.py:268]: Do we still need this whole patch for newer version megatron?

Comment (HollowMan6) [reply verl/models/mcore/patch.py:268]: Yes, as unfortunately https://github.com/NVIDIA/TransformerEngine/pull/2629 hasn't been merged into TE.

Comment (HollowMan6) [reply verl/models/mcore/patch.py:268]: and neither https://github.com/NVIDIA/Megatron-LM/pull/3003 for mcore, if one of them gets merged, then we don't need this patch to support flash attention for MLA.

File: verl/models/mcore/patch.py [hunk 2]
```diff
@@ -258,13 +258,17 @@ def patch_forward(
...
+        q_compressed = None
+        if len(qkv) > 3:
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/models/mcore/patch.py]: a mistake, fixed

File: verl/models/mcore/patch.py [hunk 3]
```diff
@@ -298,13 +302,20 @@ def patch_forward(
                 query, key, value, attention_mask, packed_seq_params=packed_seq_params
             )
         else:
+            extra_kwargs = {}
+            if getattr(self.config, "experimental_attention_variant", None) == "dsa":
+                # For dsa we need to pass in the original hidden states and the compressed
+                # query representation.
+                extra_kwargs["x"] = hidden_states
+                extra_kwargs["qr"] = q_compressed
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/models/mcore/patch.py:316]: They are added together in https://github.com/NVIDIA/Megatron-LM/commit/382eeea3951a6a333e023689e4b9e2f568f23e41#diff-d7e453d550e7f78221ea3445018264015c1e83c0db892a41253bde71d069a202L247 so this shouldn't be needed

File: verl/models/mcore/patch.py (hunk 1 above)
Comment (ISEEKYAN) [reply verl/models/mcore/patch.py:268]: I think we should find a way to remove this patch, let me check because I am currently WIP with DS3.2 supports

Comment (ISEEKYAN) [reply verl/models/mcore/patch.py:268]: this patch is originally proposed to correct the `get_query_key_value_tensors` in `multi_latent_attention.py` when packed_seq_params is not None, not for flash attention in MLA. We could remove this patch and remind the user to upgrade mcore/TE if some user is asking for FA in MLA

Comment (HollowMan6) [reply verl/models/mcore/patch.py:268]: > We could remove this patch and remind the user to upgrade mcore/TE if some user is asking for FA in MLA

Yeah, but I guess we will need to get https://github.com/NVIDIA/TransformerEngine/pull/2629 merged into TE or https://github.com/NVIDIA/Megatron-LM/pull/3003 for mcore and wait them to be released first before we proceed to remove the patch.

Comment (ISEEKYAN) [reply verl/models/mcore/patch.py:268]: that would be very long, let's just merge this anyway

**Comment by ISEEKYAN** [comment]
please fix the sanity check CI before merging it.

**Comment by HollowMan6** [comment]
> please fix the sanity check CI before merging it.

That one is from https://github.com/verl-project/verl/pull/5466#issuecomment-3991100222, but yeah I can address it in this PR if this is going to be merged now.

**Comment by ISEEKYAN** [comment]
> > please fix the sanity check CI before merging it.
> 
> That one is from [#5466 (comment)](https://github.com/verl-project/verl/pull/5466#issuecomment-3991100222), but yeah I can address it in this PR if this is going to be merged now.

soooooo quick again!
